services/src/main/java/org/keycloak/services/resources/account/LinkedAccountsBean.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2018 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.services.resources.account;
+
+import org.keycloak.models.FederatedIdentityModel;
+import org.keycloak.models.IdentityProviderModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.utils.KeycloakModelUtils;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+import org.keycloak.credential.CredentialModel;
+
+/**
+ * @author Stan Silvert
+ */
+public class LinkedAccountsBean {
+
+    private final List<LinkedAccountEntry> linkedAccounts;
+    private final boolean hasPasswordCredential;
+
+    // this private constructor used by tests during deserialization
+    private LinkedAccountsBean() {
+        this.linkedAccounts = null;
+        this.hasPasswordCredential = false;
+    }
+
+    public LinkedAccountsBean(KeycloakSession session, RealmModel realm, UserModel user) {
+        List<IdentityProviderModel> identityProviders = realm.getIdentityProviders();
+        Set<FederatedIdentityModel> identities = session.users().getFederatedIdentities(user, realm);
+
+        Set<LinkedAccountEntry> orderedSet = new TreeSet<>(LinkedAccountComparator.INSTANCE);
+        int availableIdentities = 0;
+        if (identityProviders != null && !identityProviders.isEmpty()) {
+            for (IdentityProviderModel provider : identityProviders) {
+                if (!provider.isEnabled()) {
+                    continue;
+                }
+                String providerId = provider.getAlias();
+
+                FederatedIdentityModel identity = getIdentity(identities, providerId);
+
+                if (identity != null) {
+                    availableIdentities++;
+                }
+
+                String displayName = KeycloakModelUtils.getIdentityProviderDisplayName(session, provider);
+                String guiOrder = provider.getConfig() != null ? provider.getConfig().get("guiOrder") : null;
+                LinkedAccountEntry entry = new LinkedAccountEntry(identity, displayName, provider.getAlias(), provider.getAlias(), guiOrder);
```

Comment:
`null` identities are added to the set

Reply:
I believe this is correct as-is.  The LinkedAccountEntry deals with null identities.  This bean is a refactored version of AccountFederatedIdentityBean from the existing account management console.  There, null identities are also allowed.

Thanks for the review.